File: api/admin/fix_swagger_tags.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
修复 swagger.yaml 文件，添加 3 级 tags 以支持 Apifox 目录分组
格式：RPC名称/分类/对象
"""

import yaml
import re
from pathlib import Path

# 定义路径前缀和对应的 3 级 tags 映射
# 格式：RPC名称/分类/对象
PATH_TAG_MAPPING = [
    # Admin RPC
    ("/api/admin/auth", "Admin/会话管理/认证"),
    ("/api/admin/system/user", "Admin/系统管理/用户"),
    ("/api/admin/system/role", "Admin/系统管理/角色"),
    ("/api/admin/system/menu", "Admin/系统管理/菜单"),
    ("/api/admin/system/config", "Admin/系统管理/配置"),
    ("/api/admin/system/audit", "Admin/系统管理/审计"),
    ("/api/admin/data/country", "Admin/数据管理/国家"),
    ("/api/admin/dashboard", "Admin/仪表盘/统计"),

    # IAM RPC - 用户管理
    ("/api/iam/user/user-invite", "IAM/用户管理/邀请"),
    ("/api/iam/user/{userId}/profile", "IAM/用户管理/资料"),
    ("/api/iam/user/{userId}/email", "IAM/用户管理/邮箱"),
    ("/api/iam/user/{userId}/credential", "IAM/用户管理/凭证"),
    ("/api/iam/user/{userId}/session", "IAM/用户管理/会话"),
    ("/api/iam/user/{userId}/country", "IAM/用户管理/国家授权"),
    ("/api/iam/user/user", "IAM/用户管理/代理"),

    # FAMS RPC - 用户钱包
    ("/api/fams/user/wallet", "FAMS/用户钱包/钱包"),

    # FAMS RPC - 银行管理
    ("/api/fams/bank/customer", "FAMS/银行管理/客户"),
    ("/api/fams/bank/account-application", "FAMS/银行管理/开户申请"),
    ("/api/fams/bank/account", "FAMS/银行管理/账户"),
    ("/api/fams/bank/deposit", "FAMS/银行管理/存款"),
    ("/api/fams/bank/withdrawal", "FAMS/银行管理/提现"),

    # FAMS RPC - 代理收益
    ("/api/fams/agent/earnings", "FAMS/代理管理/收益"),

    # FAMS RPC - 报表
    ("/api/fams/report", "FAMS/报表管理/报表"),

    # Public API
    ("/api/public/admin/captcha", "Public/公开接口/管理员验证码"),
    ("/api/public/admin/login", "Public/公开接口/管理员登录"),
    ("/api/public/iam/captcha", "Public/公开接口/用户验证码"),
    ("/api/public/iam/login", "Public/公开接口/用户登录"),
    ("/api/public/iam/register", "Public/公开接口/用户注册"),
    ("/api/public/iam/email", "Public/公开接口/邮箱验证"),
]

def get_tag_for_path(path):
    """根据路径获取对应的 tag"""
    # 按最长匹配优先
    for prefix, tag in sorted(PATH_TAG_MAPPING, key=lambda x: len(x[0]), reverse=True):
        if path.startswith(prefix):
            return tag
    return "其他/未分类/未知"

def add_tags_to_swagger(swagger_file):
    """给 swagger.yaml 添加 tags"""
    print(f"读取文件: {swagger_file}")

    with open(swagger_file, 'r', encoding='utf-8') as f:
        swagger = yaml.safe_load(f)

    # 收集所有使用的 tags
    used_tags = set()

    # 给每个 path 添加 tags
    if 'paths' in swagger:
        for path, methods in swagger['paths'].items():
            tag = get_tag_for_path(path)
            used_tags.add(tag)

            for method, details in methods.items():
                if isinstance(details, dict):
                    # 添加 tags 字段
                    details['tags'] = [tag]

                    # 从 summary 或 description 获取接口说明
                    summary = details.get('summary', '')
                    if summary:
                        details['summary'] = summary

    # 在文件头部添加全局 tags 定义
    tags_list = []
    for tag in sorted(used_tags):
        tags_list.append({
            'name': tag,
            'description': tag.replace('/', ' > ')  # 在描述中用 > 表示层级
        })

    swagger['tags'] = tags_list

    # 添加更多元信息
    if 'info' not in swagger:
        swagger['info'] = {}

    swagger['info']['title'] = 'AKATM Admin API'
    swagger['info']['description'] = 'AKATM 后台管理系统 API 文档\n\n目录结构：RPC名称/分类/对象'
    swagger['info']['version'] = 'v1.0'

    # 保存修改后的文件
    backup_file = str(swagger_file) + '.backup'
    print(f"备份原文件到: {backup_file}")

    # 如果已存在备份，先删除
    if Path(backup_file).exists():
        Path(backup_file).unlink()

    Path(swagger_file).rename(backup_file)

    print(f"保存修改后的文件: {swagger_file}")
    with open(swagger_file, 'w', encoding='utf-8') as f:
        yaml.dump(swagger, f, allow_unicode=True, sort_keys=False, default_flow_style=False)

    print(f"\n完成! 共添加 {len(used_tags)} 个标签:")

    # 按层级分组显示
    tags_by_rpc = {}
    for tag in used_tags:
        parts = tag.split('/')
        rpc = parts[0] if len(parts) > 0 else '未知'
        if rpc not in tags_by_rpc:
            tags_by_rpc[rpc] = []
        tags_by_rpc[rpc].append(tag)

    for rpc in sorted(tags_by_rpc.keys()):
        print(f"\n{rpc}:")
        for tag in sorted(tags_by_rpc[rpc]):
            print(f"  {tag}")

if __name__ == '__main__':
    swagger_file = Path(__file__).parent / 'swagger.yaml'
    add_tags_to_swagger(swagger_file)

File: api/admin/fix_swagger_tags.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// APIInfo 存储API文件的信息
type APIInfo struct {
	File   string
	Group  string
	Tag    string
	Prefix string
}

// 从API文件中提取group和tags信息
func extractAPIInfo(filePath string) (*APIInfo, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	info := &APIInfo{File: filepath.Base(filePath)}
	scanner := bufio.NewScanner(file)

	// 正则表达式
	groupRe := regexp.MustCompile(`group:\s*(\w+)`)
	tagsRe := regexp.MustCompile(`tags:\s*"([^"]+)"`)
	prefixRe := regexp.MustCompile(`prefix:\s*([^\s]+)`)

	inServerBlock := false

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())

		// 检测@server块
		if strings.Contains(line, "@server(") {
			inServerBlock = true
			continue
		}

		if inServerBlock {
			// 提取group
			if matches := groupRe.FindStringSubmatch(line); matches != nil {
				info.Group = matches[1]
			}

			// 提取tags
			if matches := tagsRe.FindStringSubmatch(line); matches != nil {
				info.Tag = matches[1]
			}

			// 提取prefix
			if matches := prefixRe.FindStringSubmatch(line); matches != nil {
				info.Prefix = matches[1]
			}

			// 检测块结束
			if strings.Contains(line, ")") {
				inServerBlock = false
			}
		}
	}

	if info.Group == "" || info.Tag == "" {
		return nil, fmt.Errorf("未找到group或tags: %s", filePath)
	}

	return info, scanner.Err()
}

// 扫描所有API文件
func scanAPIFiles(dir string) (map[string]string, map[string]string, error) {
	groupToTag := make(map[string]string)
	prefixToTag := make(map[string]string)

	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// 只处理.api文件
		if !info.IsDir() && strings.HasSuffix(info.Name(), ".api") {
			apiInfo, err := extractAPIInfo(path)
			if err != nil {
				fmt.Printf("⚠️  跳过 %s: %v\n", info.Name(), err)
				return nil
			}

			groupToTag[apiInfo.Group] = apiInfo.Tag
			prefixToTag[apiInfo.Prefix] = apiInfo.Tag
			fmt.Printf("✓ %s: %s (%s) -> %s\n", info.Name(), apiInfo.Group, apiInfo.Prefix, apiInfo.Tag)
		}

		return nil
	})

	return groupToTag, prefixToTag, err
}

// 处理swagger.yaml文件
func fixSwaggerTags(swaggerFile string, prefixToTag map[string]string) error {
	fmt.Printf("\n读取 Swagger 文件: %s\n", swaggerFile)

	// 读取文件
	data, err := os.ReadFile(swaggerFile)
	if err != nil {
		return fmt.Errorf("读取文件失败: %v", err)
	}

	// 解析YAML
	var swagger map[string]interface{}
	if err := yaml.Unmarshal(data, &swagger); err != nil {
		return fmt.Errorf("解析YAML失败: %v", err)
	}

	// 收集实际使用的tags
	usedTags := make(map[string]bool)
	addedCount := 0

	// 处理paths中的tags - 根据path前缀匹配tag
	if paths, ok := swagger["paths"].(map[string]interface{}); ok {
		for pathStr, methods := range paths {
			// 查找最长匹配的prefix
			var matchedTag string
			var maxPrefixLen int

			for prefix, tag := range prefixToTag {
				if strings.HasPrefix(pathStr, prefix) && len(prefix) > maxPrefixLen {
					matchedTag = tag
					maxPrefixLen = len(prefix)
				}
			}

			// 如果找到匹配的tag，添加到所有操作中
			if matchedTag != "" {
				if methodMap, ok := methods.(map[string]interface{}); ok {
					for method, operation := range methodMap {
						// 跳过非HTTP方法的键
						if method == "parameters" {
							continue
						}

						if opMap, ok := operation.(map[string]interface{}); ok {
							// 添加tags（goctl不生成tags，所以我们直接添加）
							opMap["tags"] = []interface{}{matchedTag}
							usedTags[matchedTag] = true
							addedCount++
						}
					}
				}
			}
		}
	}

	// 创建tags定义列表
	tagsList := make([]map[string]string, 0)
	sortedTags := make([]string, 0, len(usedTags))
	for tag := range usedTags {
		sortedTags = append(sortedTags, tag)
	}
	sort.Strings(sortedTags)

	for _, tag := range sortedTags {
		tagsList = append(tagsList, map[string]string{
			"name":        tag,
			"description": tag,
		})
	}

	// 添加或更新tags定义
	swagger["tags"] = tagsList

	// 更新info信息
	if info, ok := swagger["info"].(map[string]interface{}); ok {
		info["title"] = "AKATM Admin API"
		info["description"] = "AKATM 后台管理系统 API 文档"
		if _, ok := info["version"]; !ok {
			info["version"] = "v1.0"
		}
	}

	// 备份原文件
	backupFile := swaggerFile + ".backup"
	if err := os.Rename(swaggerFile, backupFile); err == nil {
		fmt.Printf("✓ 已备份原文件到: %s\n", backupFile)
	}

	// 保存修改后的文件
	output, err := yaml.Marshal(&swagger)
	if err != nil {
		return fmt.Errorf("生成YAML失败: %v", err)
	}

	if err := os.WriteFile(swaggerFile, output, 0644); err != nil {
		return fmt.Errorf("写入文件失败: %v", err)
	}

	fmt.Printf("\n✅ 完成！\n")
	fmt.Printf("   - 添加了 %d 个接口的tags\n", addedCount)
	fmt.Printf("   - 共 %d 个分类标签\n", len(usedTags))
	fmt.Printf("\n标签列表:\n")

	// 按RPC分组显示
	tagsByPrefix := make(map[string][]string)
	for tag := range usedTags {
		prefix := strings.Split(tag, "-")[0]
		tagsByPrefix[prefix] = append(tagsByPrefix[prefix], tag)
	}

	prefixes := make([]string, 0, len(tagsByPrefix))
	for prefix := range tagsByPrefix {
		prefixes = append(prefixes, prefix)
	}
	sort.Strings(prefixes)

	for _, prefix := range prefixes {
		fmt.Printf("\n【%s】\n", prefix)
		tags := tagsByPrefix[prefix]
		sort.Strings(tags)
		for _, tag := range tags {
			fmt.Printf("  • %s\n", tag)
		}
	}

	return nil
}

func main() {
	fmt.Println("🔧 Swagger Tags 修复工具")
	fmt.Println("========================================")

	// 扫描docs目录下的所有API文件
	fmt.Println("\n📂 扫描 API 文件...")
	groupToTag, prefixToTag, err := scanAPIFiles("docs")
	if err != nil {
		fmt.Printf("❌ 扫描API文件失败: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("\n📊 找到 %d 个API定义\n", len(groupToTag))

	// 处理swagger.yaml
	fmt.Println("\n🔄 处理 Swagger 文件...")
	if err := fixSwaggerTags("swagger.yaml", prefixToTag); err != nil {
		fmt.Printf("❌ 处理失败: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("\n🎉 现在可以将 swagger.yaml 导入到 Apifox 了！")
}
